### PR TITLE
[Docs] translate migrate runner

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/migration/runner.md [hunk 1]
```diff
@@ -1,3 +1,1476 @@
 # Migrate Runner from MMCV to MMEngine
 
-Coming soon. Please refer to [chinese documentation](https://mmengine.readthedocs.io/zh_CN/latest/migration/runner.html).
+## Introduction
+
+As MMCV supports more and more deep learning tasks, and users' needs become much more complicated, we have higher requirements for the flexibility and versatility of the existing `Runner` of MMCV. Therefore, MMEngine implements a more general and flexible `Runner` based on MMCV to support more complicated training processes.
+
+The `Runner` in MMEngine expands the scope and takes on more functions. we abstracted [training loop controller (EpochBasedTrainLoop/IterBasedTrainLoop)](mmengine.runner.EpochBasedLoop), [validation loop controller ( ValLoop)](mmengine.runner.ValLoop) and [TestLoop](mmengine.runner.TestLoop) to make it more convenient for users to customize their training process.
+
+Firstly, we introduce how to migrated the entry point of training from MMCV to MMEngine, to simplify and unify the training script. Then, we'll introduce the difference of the instantiation of `Runner` between MMCV and MMEngine in detail.
```
Comment (C1rN09) [inline docs/en/migration/runner.md:9]:
```suggestion
Firstly, we will introduce how to migrate the entry point of training from MMCV to MMEngine, to simplify and unify the training script. Then, we'll introduce the difference in the instantiation of `Runner` between MMCV and MMEngine in detail.
```

File: docs/en/migration/runner.md [hunk 2]
```diff
@@ -1,3 +1,1476 @@
...
+```
+
+</div>
+  </td>
+
+</tr>
+</thead>
+</table>
+
+`Runner` in MMEngine provides with more customizable components, including training/validation/testing process and DataLoader. Therefore, the configuration file is a bit longer compared to MMCV.
```
Comment (C1rN09) [inline docs/en/migration/runner.md:271]:
```suggestion
`Runner` in MMEngine provides more customizable components, including training/validation/testing process and DataLoader. Therefore, the configuration file is a bit longer compared to MMCV.
```

File: docs/en/migration/runner.md [hunk 3]
```diff
@@ -1,3 +1,1476 @@
...
+
+`MMEngine` follows the WYSIWYG principle and reorganized the hierarchy of each component in configuration so that most of the first-level fields of configuration correspond the core components in the `Runner`, such as DataLoader, [Evaluator](../tutorials/evaluation.md), [Hook](../tutorials/hook.md), etc. The new format configuration file could help users to read and understand the core components in `Runner`, and ignore the relatively unimportant parts.
```
Comment (C1rN09) [inline docs/en/migration/runner.md:273]:
```suggestion
`MMEngine` follows the WYSIWYG principle and reorganizes the hierarchy of each component in configuration so that most of the first-level fields of configuration correspond to the core components in the `Runner`, such as DataLoader, [Evaluator](../tutorials/evaluation.md), [Hook](../tutorials/hook.md), etc. The new format configuration file could help users to read and understand the core components in `Runner`, and ignore the relatively unimportant parts.
```

File: docs/en/migration/runner.md [hunk 4]
```diff
@@ -1,3 +1,1476 @@
...
+```python
+# `apis/train.py` is removed in `mmengine`
+```
+
+</td>
+</tr>
+</thead>
+</table>
+
+Table above shows the difference between training script of MMEngine `Runner` and MMCV `Runner`. Repositories of OpenMMLab 1.x organize their own process to build `Runner`, which contributes to the large amount of redundant code. MMEngine unify and format the building process, such as setting random seed, initializing distributed environment, building DataLoader, building `Optimizer`, etc. This help the downstream repositories simplify the process to prepare the runner, and only need to configure the parameters of `Runner`.
```
Comment (C1rN09) [inline docs/en/migration/runner.md:617]:
```suggestion
Table above shows the differences between training script of MMEngine `Runner` and MMCV `Runner`. Repositories of OpenMMLab 1.x organize their own process to build `Runner`, which contributes to the large amount of redundant code. MMEngine unifies and formats the building process, such as setting random seed, initializing distributed environment, building DataLoader, building `Optimizer`, etc. This help the downstream repositories simplify the process to prepare the runner, and only need to configure the parameters of `Runner`.
```

File: docs/en/migration/runner.md [hunk 5]
```diff
@@ -1,3 +1,1476 @@
...
+
+For the downstream repositories, training script based on MMEngine Runner not only simplify the `tools/train.py`, but also can directly omit the `apis/train.py`. Similarly, we can also set random seed, initialize distributed environment by configuring the parameters of `Runner`, and do not need to implement the corresponding code.
+
+## Migrate Runner
+
+This section describes the differences in the training, validation, and testing processes between the MMCV Runner and the MMEngine Runner.
+The differences of training/validation/testing process between MMCV Runner and the MMEngine Runner are as follows:
```
Comment (C1rN09) [inline docs/en/migration/runner.md:624]:
```suggestion
This section describes the differences in the training, validation, and testing processes between the MMCV Runner and the MMEngine Runner, as follows.
```

File: docs/en/migration/runner.md [hunk 6]
```diff
@@ -1,3 +1,1476 @@
...
+setup_multi_processes(cfg)
+init_dist(cfg.launcher, **cfg.dist_params)
+model = MMDistributedDataParallel(
+    model,
+    device_ids=[int(os.environ['LOCAL_RANK'])],
+    broadcast_buffers=False,
+    find_unused_parameters=find_unused_parameters)
+```
+
+As for MMEngine, you can setup launcher by configuring `launcher` of `Runner`, and configure other process mentioned above by configuring `env_cfg`. See more information in the table below:
```
Comment (C1rN09) [inline docs/en/migration/runner.md:743]:
```suggestion
As for MMEngine, you can setup launcher by configuring `launcher` of `Runner`, and configure other items mentioned above in `env_cfg`. See more information in the table below:
```

File: docs/en/migration/runner.md [hunk 7]
```diff
@@ -1,3 +1,1476 @@
...
+    })
+```
+
+</div>
+  </td>
+</tr>
+</thead>
+</table>
+
+```{note}:
```
Comment (C1rN09) [inline docs/en/migration/runner.md:1024]:
```suggestion
```{note}
```

File: docs/en/migration/runner.md [hunk 8]
```diff
@@ -1,3 +1,1476 @@
...
+
+```python
+from math import gamma
+
+param_scheduler = dict(type='MultiStepLR', milestones=[2, 3], gamma=0.1)
+```
+
+### Prepare testing/validation components
+
+MMCV implements the validation process by `EvalHook`, and we'll not talk too much about it here. Consider validation is a common process in training, MMEngine abstract validation as two independent modules: [Evaluator](../tutorials/evaluation.md) and [ValLoop](../tutorials/runner.md). We can customize the metric or the validation process by defining a new [loop](mmengine.runner.ValLoop) or a new [metric](mmengine.evaluator.BaseMetirc).
```
Comment (C1rN09) [inline docs/en/migration/runner.md:1169]:
```suggestion
MMCV implements the validation process by `EvalHook`, and we'll not talk too much about it here. Given that validation is a common process in training, MMEngine abstract validation as two independent modules: [Evaluator](../tutorials/evaluation.md) and [ValLoop](../tutorials/runner.md). We can customize the metric or the validation process by defining a new [loop](mmengine.runner.ValLoop) or a new [metric](mmengine.evaluator.BaseMetirc).
```

File: docs/en/migration/runner.md [hunk 9]
```diff
@@ -1,3 +1,1476 @@
...
+    optimizer=optimizer,
+    work_dir=work_dir,
+    logger=logger,
+    max_epochs=4
+)
+```
+
+**Building Runner in MMEngine**
+
+Besides the arguments mentioned above, such as `env_cfg`, `optim_wrapper`, etc. The `EpochBasedRunner` and `max_epochs` appeared in the example of `MMCV` are configured in `train_cfg`:
```
Comment (C1rN09) [inline docs/en/migration/runner.md:1247]:
```suggestion
The `EpochBasedRunner` and `max_epochs` arguments in `MMCV` are moved to `train_cfg` in MMEngine. All parameters configurable in `train_cfg` are listed below:
```

File: docs/en/migration/runner.md [hunk 10]
```diff
@@ -1,3 +1,1476 @@
...
+
+Complete the process mentioned above the `Runner.__init__` and `Runner.train`
+
+```python
+runner.train()
+```
+
+### Testing process
+
+Since MMCV Runner does not integrate the test function, we need to implement the test scripts by our own.
```
Comment (C1rN09) [inline docs/en/migration/runner.md:1361]:
```suggestion
Since MMCV Runner does not integrate the test function, we need to implement the test scripts by ourselves.
```

File: docs/en/migration/runner.md [hunk 11]
```diff
@@ -1,3 +1,1476 @@
...
+
+For MMEngine Runner, as long as we configure the `test_dataloader`, `test_cfg` and `test_evaluator` for the `Runner`, we can call `Runner.test` to start the testing process.
```
Comment (C1rN09) [inline docs/en/migration/runner.md:1363]:
```suggestion
For MMEngine Runner, as long as we have configured the `test_dataloader`, `test_cfg` and `test_evaluator` for the `Runner`, we can call `Runner.test` to start the testing process.
```

File: docs/en/migration/runner.md [hunk 12]
```diff
@@ -1,3 +1,1476 @@
...
+    test_dataloader=val_dataloader,
+    test_evaluator=val_evaluator,
+    test_cfg=dict(type='TestLoop'),
+)
+runner.test()
+```
+
+### Customize training process
+
+If we want to customize a training/validation process, we need to override the `Runner.val` or `Runner.train` in a custom `Runner`. Take overriding `runner.train` as an example, suppose we need to training with the same batch twice for each iteration, we can override the `Runner.train` like this:
```
Comment (C1rN09) [inline docs/en/migration/runner.md:1412]:
```suggestion
If we want to customize a training/validation process, we need to override the `Runner.val` or `Runner.train` in a custom `Runner`. Take overriding `runner.train` as an example, suppose we need to train with the same batch twice for each iteration, we can override the `Runner.train` like this:
```

**Review by C1rN09 — APPROVED** [review]

File: docs/en/migration/runner.md [hunk 13]
```diff
@@ -1,3 +1,1475 @@
...
+</thead>
+</table>
+
+`Runner` in MMEngine provides more customizable components, including training/validation/testing process and DataLoader. Therefore, the configuration file is a bit longer compared to MMCV.
+
+`MMEngine` follows the WYSIWYG principle and reorganizes the hierarchy of each component in configuration so that most of the first-level fields of configuration correspond to the core components in the `Runner`, such as DataLoader, [Evaluator](../tutorials/evaluation.md), [Hook](../tutorials/hook.md), etc. The new format configuration file could help users to read and understand the core components in `Runner`, and ignore the relatively unimportant parts.
+
+### Migrate the training script
+
+Compared with the `Runner` in MMCV, `Runner` in MMEngine takes on more functions, such as building Dataloader and distributed model. Therefore, we do not need to build the components like DataLoader and distributed model manually anymore. We can configure them during the instantiation of `Runner`, and then build them in the training/validation/testing process. Take the training script of MMDet as an example:
```
Comment (zhouzaida) [inline docs/en/migration/runner.md:277]:
```suggestion
Compared with the `Runner` in MMCV, `Runner` in MMEngine takes on more functions, such as building DataLoader and distributed model. Therefore, we do not need to build the components like DataLoader and distributed model manually anymore. We can configure them during the instantiation of `Runner`, and then build them in the training/validation/testing process. Take the training script of MMDet as an example:
```

File: docs/en/migration/runner.md [hunk 14]
```diff
@@ -1,3 +1,1475 @@
...
+            paramwise_cfg=paramwise_cfg))
+    optimizer = optim_constructor(model)
+    return optimizer
+
+optimizer = build_optimizer(model, optimizer_cfg)
+```
+
+**Prepare optimizer in MMEngine**
+
+MMEngine needs to configure [optim_wrapper](mmengine.optim.OptimWrapper) for `Runner`. For more complicated cases, you can also configure the `optim_wrapper` more specifically. See more information in the api [documents](mmengine.runner.Runner.build_optim_wrapper)
```
Comment (zhouzaida) [inline docs/en/migration/runner.md:967]:
```suggestion
MMEngine needs to configure [optim_wrapper](mmengine.optim.OptimWrapper) for `Runner`. For more complicated cases, you can also configure the `optim_wrapper` more specifically. See more information in the API [documents](mmengine.runner.Runner.build_optim_wrapper)
```

File: docs/en/migration/runner.md [hunk 15]
```diff
@@ -1,3 +1,1475 @@
...
+
+```python
+from math import gamma
+
+param_scheduler = dict(type='MultiStepLR', milestones=[2, 3], gamma=0.1)
+```
+
+### Prepare testing/validation components
+
+MMCV implements the validation process by `EvalHook`, and we'll not talk too much about it here. Given that validation is a common process in training, MMEngine abstract validation as two independent modules: [Evaluator](../tutorials/evaluation.md) and [ValLoop](../tutorials/runner.md). We can customize the metric or the validation process by defining a new [loop](mmengine.runner.ValLoop) or a new [metric](mmengine.evaluator.BaseMetirc).
```
Comment (zhouzaida) [inline docs/en/migration/runner.md:1168]:
```suggestion
MMCV implements the validation process by `EvalHook`, and we'll not talk too much about it here. Given that validation is a common process in training, MMEngine abstracts validation as two independent modules: [Evaluator](../tutorials/evaluation.md) and [ValLoop](../tutorials/runner.md). We can customize the metric or the validation process by defining a new [loop](mmengine.runner.ValLoop) or a new [metric](mmengine.evaluator.BaseMetirc).
```

**Review by zhouzaida — APPROVED** [review]
